File: drills/elixir/solutions/nc72_maximum_product_subarray.exs
defmodule Solution do
  def max_product([]), do: 0

  # A negative number turns the best running product into the worst and the
  # worst into the best, so both have to be carried. Zero resets them both,
  # which falls out of taking the element itself as an option.
  def max_product([first | rest]) do
    {_high, _low, best} =
      Enum.reduce(rest, {first, first, first}, fn n, {high, low, best} ->
        candidates = [n, high * n, low * n]
        high = Enum.max(candidates)
        {high, Enum.min(candidates), max(best, high)}
      end)

    best
  end
end

File: drills/go/solutions/nc72_maximum_product_subarray.go
package main

func maxProduct(nums []int) int {
	// A negative flips the largest and smallest products ending here, so
	// carry both: the smallest may become the largest at the next negative.
	best := nums[0]
	largest, smallest := 1, 1
	for _, n := range nums {
		candidates := []int{n, largest * n, smallest * n}
		largest = max(candidates[0], max(candidates[1], candidates[2]))
		smallest = min(candidates[0], min(candidates[1], candidates[2]))
		best = max(best, largest)
	}
	return best
}
